pocs/camera/libasi.py

```diff
@@ -0,0 +1,597 @@
+        video_data = self._image_array(width, height, image_type)
+        timeout = int(get_quantity_value(timeout, unit=u.ms))
+        try:
+            self._call_function('ASIGetVideoData',
+                                camera_ID,
+                                video_data.ctypes.data_as(ctypes.POINTER(ctypes.c_byte)),
+                                ctypes.c_long(video_data.nbytes),
+                                ctypes.c_int(-1))
+            # If set timeout to anything but -1 (no timeout) this call times out instantly?
+        except RuntimeError:
+            # Expect some dropped frames during video capture
+            return None
+        else:
+            return video_data
+
+    # Private methods
+
+    def _call_function(self, function_name, camera_ID, *args):
+        """ Utility functions for calling the functions that take return ErrorCode """
+        function = getattr(self._CDLL, function_name)
+        error_code = function(ctypes.c_int(camera_ID), *args)
+        if error_code != ErrorCode.SUCCESS:
+            msg = "Error calling {}: {}".format(function_name, ErrorCode(error_code).name)
+            self.logger.error(msg)
+            raise RuntimeError(msg)
+
+    def _parse_info(self, camera_info):
+        """ Utility function to parse CameraInfo Structures into something more Pythonic """
+        pythonic_info = {'name': camera_info.name.decode(),
+                         'camera_ID': int(camera_info.camera_ID),
+                         'max_height': camera_info.max_height * u.pixel,
+                         'max_width': camera_info.max_width * u.pixel,
+                         'is_colour_camera': bool(camera_info.is_color_camera),
+                         'bayer_pattern': BayerPattern(camera_info.bayer_pattern).name,
+                         'supported_bins': self._parse_bins(camera_info.supported_bins),
+                         'supported_video_format': self._parse_formats(
+                             camera_info.supported_video_format),
+                         'pixel_size': camera_info.pixel_size * u.um,
+                         'has_mechanical_shutter': bool(camera_info.has_mechanical_shutter),
+                         'has_ST4_port': bool(camera_info.has_ST4_port),
+                         'has_cooler': bool(camera_info.has_cooler),
+                         'is_USB3_host': bool(camera_info.is_USB3_host),
+                         'is_USB3_camera': bool(camera_info.is_USB3_camera),
+                         'e_per_adu': camera_info.e_per_adu * u.electron / u.adu,
+                         'bit_depth': camera_info.bit_depth * u.bit,
+                         'is_trigger_camera': bool(camera_info.is_trigger_camera)}
+        return pythonic_info
+
+    def _parse_bins(self, supported_bins):
+        bins = tuple(int(bin) for bin in supported_bins if bin != 0)
+        return bins
+
+    def _parse_formats(self, supported_formats):
+        formats = []
+        for i in range(8):
```

Comment:
It could change, but only if the API of the SDK changes. The `supported_video_formats` field of the camera info C `struct` is an array of 8 `int`s. There's no real reason not to use `len(supported_formats)` here, though.

Reply:
Better yet, `for supported_format in supported_formats:`